libafl/src/stages/mod.rs

```diff
@@ -122,51 +111,75 @@ where
     E: UsesState<State = S>,
     EM: UsesState<State = S>,
     Z: UsesState<State = S>,
-    S: UsesInput,
+    S: UsesInput + HasStageStatus,
 {
     fn perform_all(
         &mut self,
         _: &mut Z,
         _: &mut E,
-        _: &mut S,
+        stage: &mut S,
         _: &mut EM,
-        _: CorpusId,
     ) -> Result<(), Error> {
-        Ok(())
+        if stage.current_stage()?.is_some() {
+            Err(Error::illegal_state(
+                "Got to the end of the tuple without completing resume.",
+            ))
+        } else {
+            Ok(())
+        }
     }
 }
 
 impl<Head, Tail, E, EM, Z> StagesTuple<E, EM, Head::State, Z> for (Head, Tail)
 where
     Head: Stage<E, EM, Z>,
-    Tail: StagesTuple<E, EM, Head::State, Z>,
+    Tail: StagesTuple<E, EM, Head::State, Z> + HasConstLen,
     E: UsesState<State = Head::State>,
     EM: UsesState<State = Head::State>,
     Z: UsesState<State = Head::State>,
+    Head::State: HasStageStatus,
 {
     fn perform_all(
         &mut self,
         fuzzer: &mut Z,
         executor: &mut E,
         state: &mut Head::State,
         manager: &mut EM,
-        corpus_idx: CorpusId,
     ) -> Result<(), Error> {
-        // Perform the current stage
-        self.0
-            .perform(fuzzer, executor, state, manager, corpus_idx)?;
+        match state.current_stage()? {
+            Some(idx) if idx < Self::LEN => {
+                // do nothing; we are resuming
+            }
+            Some(idx) if idx == Self::LEN => {
+                // perform the stage, but don't set it
+                Head::Progress::initialize_progress(state)?;
+                self.0.perform(fuzzer, executor, state, manager)?;
+                Head::Progress::clear_progress(state)?;
+                state.clear_stage()?;
+            }
+            Some(idx) if idx > Self::LEN => {
+                unreachable!("We should clear the stage index before we get here...");
+            }
+            // this is None, but the match can't deduce that
+            _ => {
```

Comment:
when is state.current_stage() None? when using StdState?
It always return Ok() right?

Reply:
This None branch is when you are normally executing stuff (not resuming)?

Reply:
ok. but https://github.com/AFLplusplus/LibAFL/pull/1780/files#diff-a5d4da7068b0cb9792a4e92f3e7a8a3403323466332f7de67f7211ae9b9afbf2R564 says it never returns None

Reply:
It definitely can. `.get` returns an Option, so if we are not currently in a stage (e.g. if the stage depth exceeds the current size of the Vec, such as when we've just entered an inner stage and are not resuming) it will return None.

Reply:
ah sorry i was confused